packages/dart_firebase_admin/lib/src/google_cloud_firestore/firestore.dart

```diff
@@ -193,6 +196,79 @@ class Firestore {
 
     return reader.get(tag);
   }
+
+  Future<T> runTransaction<T>(
+    TransactionHandler<T> transactionHandler, {
+    int maxAttempts = 5,
+    Duration timeout = const Duration(seconds: 30),
+  }) async {
+    if (maxAttempts <= 0) {
+      return Future.error(
+        FirebaseFirestoreAdminException(
+          FirestoreClientErrorCode.failedPrecondition,
+          'Transaction max attemps must be > 0.',
+        ),
+      );
+    }
+    return Future<T>(() async {
+      var attempts = 1;
+      while (true) {
+        try {
+          return await _runTransaction(transactionHandler);
+        } on FirebaseFirestoreAdminException catch (firebaseError) {
+          if (firebaseError.errorCode == FirestoreClientErrorCode.aborted ||
+              firebaseError.errorCode == FirestoreClientErrorCode.unavailable ||
+              firebaseError.errorCode == FirestoreClientErrorCode.internal) {
+            if (attempts < maxAttempts) {
+              attempts += 1;
+            } else {
+              return Future.error(
+                FirebaseFirestoreAdminException(
+                  FirestoreClientErrorCode.failedPrecondition,
+                  'Transaction max atempts exceded.',
+                ),
+              );
+            }
+          } else {
+            return Future.error(firebaseError);
+          }
+        } catch (e) {
+          return Future.error(e);
+        }
+      }
+    }).timeout(
+      timeout,
+      onTimeout: () => Future.error(
+        FirebaseFirestoreAdminException(
+          FirestoreClientErrorCode.deadlineExceeded,
+          'Transaction timout.',
+        ),
+      ),
+    );
+  }
+
+  Future<T> _runTransaction<T>(
+    TransactionHandler<T> transactionHandler,
+  ) async {
+    final transactionRequest = firestore1.BeginTransactionRequest();
+
+    return _client.v1(
+      (client) async {
+        final transactionResponse = await client.projects.databases.documents
+            .beginTransaction(transactionRequest, _formattedDatabaseName)
+            .catchError(_handleException);
+
+        final transactionId = transactionResponse.transaction;
+        final transaction = Transaction(this, transactionId!);
+        final result = await transactionHandler(transaction);
+        await transaction._transactionWriteBatch
```

Comment:
Would it make sense to either have a getter for `transaction._transactionWriteBatch` or implement `commit()` and `catchError()` as part of `Transaction`? This would avoid exposing a private member (`_transactionWriteBatch`).